keps/sig-api-machinery/20190621-apply-scalability.md

```diff
@@ -0,0 +1,154 @@
+
+| Scenario | LIST | PATCH | LIST while PATCH |
+|---|---|---|---|
+| 1. |360.4 ms | 8.0 ms | 361.0 ms |
+| 2. | 1294.5 ms | 38.5 ms | 2529.6 ms |
+| 3. | 480.8 ms | 11.3 ms | 500.2 ms |
+| 4. | 1260.1 ms | 9.7 ms | 1503.2 ms |
+
+### Goals
+
+TODO
+
+### Non-Goals
+
+TODO
+
+## Proposal
+
+We have considered a few overlapping solutions:
+
+### Only send managedFields if requested:
+
+Add a query parameter (or some other mechanism) to allow users to request to see the managedFields field in metadata. If the parameter is missing,  strip the field on reads and writes (in responses). The benefit here is we can skip serializing the field most of the time. Most clients won’t and shouldn’t have to understand how to read this field. So making them receive it is a waste of resources, and confusing.
+
+### Store managedFields as a blob
+
+Instead of a highly structured nested map, with lots of small fields, we could store it as a blob. The blob would save on allocation/GC of small objects when deserializing. We could also gzip the entire thing to save space. The downside is that it would make the field harder/impossible for a user to understand. This is mitigated by combining with solution 1, or by client-go/kubectl unzipping the field automatically.
+
+### Convert managedFields directly to internal fieldpath
+
+On updates and applies, the apiserver needs to convert managedFields into an internal representation. We originally decoupled the managedFields API from the internal representation of a fieldset, to allow arbitrary changes to the internal representation without breaking the API. However converting from a serialized config to a gostruct and then into a different struct causes a lot of allocations of small objects and stress on the garbage collector. Instead, we can directly serialize as the internal representation.
+
+The internal format could be stored in protobuf, which would make it much smaller than json (json has a lot of repeated fields that make it extremely large).
+
+### Only update managedFields if previously applied to
+
+Most objects created with post or put are never applied to, so tracking managedFields on them is a waste of resources. If an object without managedFields does get applied to, we can still generate a single managedFields entry for all the existing fields. The generated managedFields conflicts set will be the same, but we won’t have any way of knowing know who is being conflicted with. CPU and memory cost on the apiserver will be dependent on the number of objects managed by server-side apply. It would increase more gradually as controllers decided to adopt SSA. Adoption could be decided on a controller by controller basis, weighing the cost vs. benefits each time. If someone applies to an object created by post or put, the conflicts they see would not contain information about the timestamp or field manager, so conflicts become less useful.
+
+### Storing in one format, serving in another
```

Comment:
Whether we do it as a subresource or not, this is the favorite solution because it significantly increases the speed of all operations, keeps a "kubernetes"-style API and reduces the etcd impact. Main drawback is that we probably need two fields: 1 for the storage, 1 that is served.